GH-2883: Fix for an NPE in OpAsQuery.

Aklakan opened the pull request:
GitHub issue resolved #2883 

Pull request Description: Changed a line to pass on an ExprTransform rather than null.

Notes for possible further improvements outside the scope of this PR:
* It may be the case that `ElementTransformer` needs to create an identity ExprTransform rather than passing null into the machinery.
* `QueryTransformOps.transformVarExprList` in my IDE warns that the `changed` flag is not used. Perhaps this needlessly creates clones of queries w.r.t. identity transforms.

----

 - [x] Tests are included.
 - [x] Commits have been squashed to remove intermediate development commit messages.
 - [x] Key commit messages start with the issue number (GH-xxxx)

By submitting this pull request, I acknowledge that I am making a contribution to the Apache Software Foundation under the terms and conditions of the [Contributor's Agreement](https://www.apache.org/licenses/contributor-agreements.html).

----

See the [Apache Jena "Contributing" guide](https://github.com/apache/jena/blob/main/CONTRIBUTING.md).

Aklakan commented on jena-arq/src/main/java/org/apache/jena/sparql/syntax/syntaxtransform/ExprTransformApplyElementTransform.java:
```diff
@@ -43,7 +43,7 @@ public ExprTransformApplyElementTransform(ElementTransform transform, boolean al
     public Expr transform(ExprFunctionOp funcOp, ExprList args, Op opArg)
     {
-        Element el2 = ElementTransformer.transform(funcOp.getElement(), transform);
+        Element el2 = ElementTransformer.transform(funcOp.getElement(), transform, this);
```
This is my actual fix attempt.

Aklakan replied:
The fix is extended with a few additional line in a separate comment. Closing this one.

Aklakan commented on jena-arq/src/main/java/org/apache/jena/sparql/syntax/syntaxtransform/QueryTransformOps.java:
```diff
@@ -69,7 +69,7 @@ public static Query transform(Query query, ElementTransform transform, ExprTrans
         // Mutate the q2 structures which are already allocated and no other code can access yet.
 
-        mutateByQueryType(q2, transform, exprTransform);
+        mutateByQueryType(q2, exprTransform);
```
Removed unused argument.

afs commented on jena-arq/src/test/java/org/apache/jena/sparql/algebra/TestOpAsQuery.java:
```diff
@@ -482,6 +482,21 @@ public void testMinus02() {
                                                            "SELECT * { GRAPH ?g { { ?x ?y ?z FILTER(EXISTS { ?s ?p ?o }) } ?x ?y ?z } }",
                                                               syntaxARQ); }
 
+    @Test
```
These EXISTS tests are about the ARQ extensions.

IIRC It's not about EXISTS, it's about subquery select expressions.

If that's right please use this test (insert at around L463):

```java
  @Test
  public void testSubQuery01() {
      test_roundTripQuery("SELECT ?z { SELECT ?x {  } }");
  }
```

The outer `SELECT ?z` stops the outer `SELECT` getting removed because `SELECT *` is a no-op.

Aklakan replied:
Your query does not fail for me.

The fix is in 
```
    public Expr transform(ExprFunctionOp funcOp, ExprList args, Op opArg)
    {
        Element el2 = ElementTransformer.transform(funcOp.getElement(), transform, this);
    }
```
So the issue is related to (NOT) EXISTS *within* in a subquery.
Also note that, the transform is an `ElementTransform` - so `SELECT * { ?s a ?c }` is different from `?s a ?c` - although the Op would be the same.

I reduced the query to this:
```
SELECT ?s {
  SELECT ?s {
    ?s a ?t
    FILTER EXISTS { SELECT * { ?s a ?c } }
  }
}
```

Aklakan commented on jena-arq/src/test/java/org/apache/jena/sparql/algebra/TestOpAsQuery.java:
```diff
@@ -329,6 +329,11 @@ public void testSubQuery3() {
     }
 
+    @Test
+    public void testSubQuery4() {
```
I added this test case for completeness - but it did not fail.

Aklakan commented:
My change is in here:

```sparql
    public Expr transform(ExprFunctionOp funcOp, ExprList args, Op opArg)
    {
        Element el2 = ElementTransformer.transform(funcOp.getElement(), transform, this);
    }
```

So the issue is related to (NOT) EXISTS when there is a subquery.
Also note that, the transform is an ElementTransform - so SELECT * { ?s a ?c } is treated different than ?s a ?c - although the Op would be the same.

I reduced the query to this:

```sparql
SELECT ?x {
  SELECT ?x { # <-- NPE disappears when removing this sub select.
    ?x a ?y
    FILTER EXISTS { SELECT * { ?y a ?z } }
  }
}
```

I silghtly updated the test cases.

Aklakan commented:
I converted to draft for now because it seems there are some more issues related to how ExprFunctionOp are handled - I added another example query that breaks to https://github.com/apache/jena/issues/2883#issuecomment-2527919893

Aklakan commented:
I updated the code that handles ExprFunctionOp that when the Element is absent then it will be derived from the algebra.

Aklakan commented on jena-arq/src/main/java/org/apache/jena/sparql/syntax/syntaxtransform/ExprTransformApplyElementTransform.java:
```diff
@@ -43,7 +44,15 @@ public ExprTransformApplyElementTransform(ElementTransform transform, boolean al
     public Expr transform(ExprFunctionOp funcOp, ExprList args, Op opArg)
     {
-        Element el2 = ElementTransformer.transform(funcOp.getElement(), transform);
+        // If the element is null then an attempt is made to obtain it from the algebra.
+        // A given element takes precedence over the algebra.
+        Element el1 = funcOp.getElement();
+        if (el1 == null) {
+            Op op = funcOp.getGraphPattern();
+            el1 = op == null ? null : OpAsQuery.asElement(op);
+        }
+        // ElementTransformer will warn should it happen that null is passed to it.
+        Element el2 = ElementTransformer.transform(el1, transform, this);
```
This is now the updated actual fix - together with the introduction of `OpAsQuery.asElement`.
